### ISSUE TITLE
Unable to unlock database after update to 2.2.0.6

### ISSUE THREAD

**Issue by valgrind** [bpellin/keepassdroid#242]
I see the app updated to 2.2.0.6 on my Pixel 2 14h ago, and today I can no longer unlock ("Invalid password or key file") the database. Since I don't see 20 open I issues about this I assume this is not a widespread problem.

Any idea what has gone wrong here?

**Comment by valgrind** [comment]
Which version was on the app store before the 2.2.0.6 update? I could side-load that and see if the problem goes away after a downgrade.

**Comment by alphasmerdon** [comment]
I also have the same issue on a Nexus 5X and 6P - Oreo. It seems to happen after you write to the database, exit, then try to log in again.

**Comment by bpellin** [comment]
I wasn't able to replicate this. Is anyone able to make a new database (without private information) that experiences this issue and post the before and after .kdbx to this page along with the password

**Comment by sophom** [comment]
I can confirm this. Here is how to reproduce the problem

1) create new database on PC using keepassx.
2) add single entry to db
3) copy file to android phone
4) open with keepassdroid (this works fine)
5) add another entry to the db in keepassdroid
6) exit keepassdroid. Make sure it stops running. You might have to kill it 
7) try opening the db again.. and "Invalid password..."

BTW the corrupted file can't be opened with keepassdroid or keepassx

I'm attaching a gz with two files. The original, which you can open and the corrupted one. The password is "pa455word"

[test_kbdx.tar.gz](https://github.com/bpellin/keepassdroid/files/1502447/test_kbdx.tar.gz)

**Comment by sophom** [comment]
Oops "p455word"

**Comment by bpellin** [comment]
Thanks a lot sophom. You are affected by this issue if you use .kdbx less than version 4 of the format and if you use a number of AES encryption rounds not equal to the KeePassDroid default of 6000.

If your DB uses 100,000 encryption rounds, we store 100,000 rounds in the database header, however, we always use 6000 rounds when computing the encryption key regardless of what was specified. I'll get an update pushed out tonight to correct this.

**Comment by bpellin** [comment]
Fixed in 2.2.0.7.

**Comment by sophom** [comment]
I can confirm the fix in 2.2.0.7. I verified that I can modify an existing kdbx file, exit keepassdroid and then reopen the file. I also copied the modified file to a PC and verified that I can open it with keepassx. 

I also tried opening the previously corrupted file and I was still not able to. That's probably OK so long as nobody corrupted their old file and have no backup.

**Comment by c1chy** [comment]
Hi,

Unfortunately I am affected by this bug after I added entry to my DB in 2.2.0.6 and am no longer able to open DB on Android nor PC (I have set 1000000 rounds for AES).

I do have a backup but it's few months old and is missing few new important entries. Is there any way I can recover access to my DB?

Thank you.
Greg

**Comment by bpellin** [comment]
I mis-characterized the bug in my last post. Version 2.2.0.6 will use the correct number of database rounds to build the key but always mark 6000 rounds in the header. So, in order for this to be correctable, you must know how many encryption rounds your database was using previously.

In 2.2.0.8, I have added a new fix mode. Under the Settings > Application > "Key encryption rounds before corruption" option you can enter the number of encryption rounds used by your database before it was corrupted (defaulting to 100,000). If decryption fails due to an incorrect password, we will retry decryption using the number of rounds you specify here. If you input the correct number your database will open. If you make a changed to your database and save it, you will get a correct version of the database stored.

**Comment by sophom** [comment]
Thank you very much bpellin! This will be appreciated by many. I can confirm that when using 2.2.0.8 I can open a previously corrupted kdbx file and save it "uncorrupted" after I set the correct number of rounds in "Key encryption rounds before corruption". To find the encrypted rounds I used keepassx   Database->Database settings.  Unfortunately that implies that I had an uncorrupted version of the file to begin with. I suppose if people don't know the number of rounds they can try a few "defaults" used by some keepass compatible programs now and in the past such as 50000 and 100000,
